src/app/routes/api/v3/admin/forms/__tests__/admin-forms.twilio.routes.spec.ts

```diff
@@ -0,0 +1,341 @@
+      await logoutSession(request)
+
+      const response = await session
+        .put(`/admin/forms/${formToUpdate._id}/twilio`)
+        .send(TWILIO_CREDENTIALS)
+
+      expect(response.status).toEqual(401)
+      expect(response.body).toEqual({ message: 'User is unauthorized.' })
+    })
+
+    it('should return 403 when user does not have permissions to update form', async () => {
+      const { user } = await dbHandler.insertFormCollectionReqs()
+      const session = await createAuthedSession(user.email, request)
+
+      // Create separate user
+      const collabUser = (
+        await dbHandler.insertFormCollectionReqs({
+          userId: new ObjectId(),
+          mailName: 'collab-user',
+          shortName: 'collabUser',
+        })
+      ).user
+
+      const randomForm = await FormModel.create({
+        title: 'form that user has no write access to',
+        admin: collabUser._id,
+        publicKey: 'some random key',
+        // Current user only has read access.
+        permissionList: [{ email: user.email }],
+        _id: MOCK_FORM_ID,
+      })
+
+      const response = await session
+        .put(`/admin/forms/${randomForm._id}/twilio`)
+        .send(TWILIO_CREDENTIALS)
+
+      expect(response.status).toEqual(403)
+      expect(response.body).toEqual({
+        message: `User ${user.email} not authorized to perform write operation on Form ${randomForm._id} with title: ${randomForm.title}.`,
+      })
+    })
+
+    it('should return 404 when form to update cannot be found', async () => {
+      const { user } = await dbHandler.insertFormCollectionReqs()
+      const session = await createAuthedSession(user.email, request)
+      const invalidFormId = MOCK_FORM_ID
+
+      const response = await session
+        .put(`/admin/forms/${invalidFormId}/twilio`)
+        .send(TWILIO_CREDENTIALS)
+
+      expect(response.status).toEqual(404)
+      expect(response.body).toEqual({ message: 'Form not found' })
+    })
+
+    it('should return 422 when id of user adding twilio credentials is not found', async () => {
+      const { form: formToUpdate, user } = await dbHandler.insertEmailForm()
+      const session = await createAuthedSession(user.email, request)
+
+      // Delete user after login.
+      await dbHandler.clearCollection(UserModel.collection.name)
+
+      const response = await session
+        .put(`/admin/forms/${formToUpdate._id}/twilio`)
+        .send(TWILIO_CREDENTIALS)
+
+      expect(response.status).toEqual(422)
+      expect(response.body).toEqual({ message: 'User not found' })
+    })
+
+    it('should return 500 when application error occurs whilst updating credentials', async () => {
+      const { form: formToUpdate, user } = await dbHandler.insertEmailForm()
+      const session = await createAuthedSession(user.email, request)
+
+      jest
+        .spyOn(AdminFormService, 'createTwilioCredentials')
+        .mockReturnValueOnce(errAsync(new DatabaseError()))
```

Comment:
this seems to be misnamed, should this be `ApplicationError` instead?

Reply:
Typo for description of test case XD